File: java/dsa/data-structures/stack-queue/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] n1, int[] n2) {
        Map<Integer, Integer> map = new HashMap<>(); // @viz: init_map
        Stack<Integer> st = new Stack<>(); // @viz: init_stack
        
        for (int val : n2) { // @viz: loop
            while (!st.isEmpty() && val > st.peek()) { // @viz: while_check
                map.put(st.pop(), val); // @viz: map_set
            }
            st.push(val); // @viz: push
        }
        
        int[] res = new int[n1.length];
        for (int i = 0; i < n1.length; i++) {
            res[i] = map.getOrDefault(n1[i], -1);
        }
        return res; // @viz: finish
    }
}

File: javascript/dsa/data-structures/stack-queue/next-greater-element-i.js
function nextGreaterElement(nums1, nums2) {
  const resultStorage = new Map(); // @viz: init_map
  const monoStack = []; // @viz: init_stack

  for (let num of nums2) { // @viz: loop
    while (monoStack.length > 0 && num > monoStack[monoStack.length - 1]) { // @viz: while_check
      const topVal = monoStack.pop(); // @viz: pop_stack
      resultStorage.set(topVal, num); // @viz: map_set
    }
    monoStack.push(num); // @viz: push
  }

  return nums1.map(n => resultStorage.get(n) ?? -1); // @viz: finish
}